### ISSUE TITLE
delay()

### ISSUE THREAD

**Issue by TheNarcissisticArsonist** [TheNarcissisticArsonist/Robot-LEDs-2016#3]
I can try to come up with a system using millis() and Δt, but we can't use delay() at all.

**Comment by danieljnchen** [comment]
In setup(), you set startTime equal to millis(); as setup() runs only once at the beginning of the program, it would be unnecessary to set it for use as a timer, and not useful to use it as a counter since the arduino started, as we could start it at almost any time relative to the match start. Instead, I'm only declaring startTime, as well as moving the declarations for int state, int stateRead, and long elapsedTime to the beginning, as declarations only need to be run once.

**Comment by TheNarcissisticArsonist** [comment]
If you declare state, stateRead, and elapsedTime in void setup(), they'll go out of scope. That's why I did their declarations at the top of the program.

In terms of your point about when we turn the arduino on, that's a really good point. I didn't think of that.

What if we defined a new idle state that ran when we turned it on. Then, when the match actually starts, we can send a signal in the robot auton code to switch to the game LEDs? For that matter, we could also do so at the end of the game. What do you think?

**Comment by danieljnchen** [comment]
Good point. On GitHub, they are declared in the void loop(), which declares them multiple times - I considered moving them to setup(), but since it puts them out of scope, I'll move the declarations to the top of the program.

Good idea with the idle states - I'll put the framework for that in the version I'm working on - I'm almost done fixing the issues you've brought up.

**Comment by TheNarcissisticArsonist** [comment]
I put them in void loop because this keeps them scoped, but I guess it doesn't matter too much at this point. If memory becomes a problem later, then we'll have to do something about it. But for now, putting them at the top shouldn't matter.

Just remember to commit frequently and use push/pull frequently!

**Comment by danieljnchen** [comment]
Ok, I can leave them in the loop(), assuming redundant declarations aren't a problem. Is that good?

**Comment by TheNarcissisticArsonist** [comment]
I found this article on how loop() actually works.

http://web.cecs.pdx.edu/~eas199/A/notes/06/whats_this_void_loop.pdf

I think there's actually another loop running at a lower level that just calls void loop() over and over, returning the function.

**Comment by danieljnchen** [comment]
Interesting - it appears that there is a function named loop that is not in itself the loop, but is instead automatically run in a loop. Clever.

On a different note, the variables (currentTime, elapsedTime, etc) are being used in functions, and are out of scope unless declared at the top as global. I pushed the latest version - tell me if you notice something wrong, or fix it yourself if you feel like it.

**Comment by TheNarcissisticArsonist** [comment]
I'll probably poke at it tonight since there's a snow day :)
